text/0000-dockerfiles.md

```diff
@@ -0,0 +1,218 @@
+# Meta
+[meta]: #meta
+- Name: Support Dockerfiles
+- Start Date: 2021-06-30
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks ([RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md)).
+
+# Motivation
+[motivation]: #motivation
+
+[RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md) introduces complexity by defining an API that allows buildpacks to modify base images. To avoid this complexity, we could rely on generated Dockerfiles for base image manipulation. This would simplify the original proposal by, e.g., only requiring a copy of the buildpack on the build-time base image.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with dynamically-generated build-time and runtime Dockerfiles that act as pre-build hooks.
+These hooks would participate in detection and execute before the buildpack build process.
+
+For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder that does not contain hooks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "hook" directories in `/cnb/hooks/`.
```

Comment:
Rather than introducing the "hook" concept, could we tie this to any existing concept (especially if it's something in Docker). For example, the [`FROM` instruction](https://docs.docker.com/engine/reference/builder/#from) initializes a new build stage and sets the Base Image for subsequent instructions. 

Can we tie this concept to "stage", or "base-image" or even "from"?

The other problem with "hook" is that it implies (to me) that it's some asynchronous fire-and-forget thing like a Webhook.

Reply:
After discussion in the Working Group, I would be comfortable with any of the following alternatives:
* "ext" - `/cnb/ext/`, `ext.toml`
* "from" - `/cnb/from/`, `from.toml`
* "stage" - `/cnb/stages/`, `stage.toml`

I would prefer `stage`, although I wish there was some way to make it clear that these were stages _before_ the buildpack build.

Reply:
+1 on `staging`. I'm also in favor of something like `ext` as proposed by @jabrown85 and @samj1912 .

I'm tying this in with the idea of providing some differentiator (suffix) for files shared with the existing buildpack API.

Here's an example of what the dir structure may look like for either cases.

### `staging` with suffixed file extensions

##### Pros

* Clear_ish_ about its purpose to "stage" the environment.

##### Cons

* Limiting to being only a "pre" build process. May we want "post" processing in the future?

##### Prior Art

* Loose tie to "stages" in Dockerfiles (see `FROM` and multi-stage).

```text
.
└── cnb/
    ├── staging/
    │   ├── stage-w-dir-only/
    │   │   └── 0.1.0/
    │   │       ├── launch.stage.toml
    │   │       ├── build.stage.toml
    │   │       ├── Dockerfile
    │   │       └── staging.toml
    │   └── stage-w-dynamic-dockerfile/
    │       └── 1.2.3/
    │           ├── bin/
    │           │   ├── build
    │           │   └── detect
    │           └── staging.toml
    └── buildpacks/
        └── buildpack-1/
            └── 0.1.0/
                ├── bin/
                │   ├── build
                │   └── detect
                └── buildpack.toml
```

### `ext` with suffixed file extensions

##### Pros

* Flexible to provide additional future functionality.

##### Cons

* A little ambiguous as `ext`.
* If we use `extensions`, then it may collide in terms to "[extension](https://github.com/buildpacks/spec/tree/main/extensions)" specifications.

##### Prior Art

* Kernel extensions ([kext](https://support.apple.com/guide/deployment-reference-macos/kernel-extensions-in-macos-apd37565d329/web))

```text
.
└── cnb/
    ├── ext/
    │   ├── ext-w-dir-only/
    │   │   └── 0.1.0/
    │   │       ├── launch.ext.toml
    │   │       ├── build.ext.toml
    │   │       ├── Dockerfile
    │   │       └── ext.toml
    │   └── ext-w-dynamic-dockerfile/
    │       └── 1.2.3/
    │           ├── bin/
    │           │   ├── build
    │           │   └── detect
    │           └── ext.toml
    └── buildpacks/
        └── buildpack-1/
            └── 0.1.0/
                ├── bin/
                │   ├── build
                │   └── detect
                └── buildpack.toml
```

Reply:
I like /cnb/stages/, stage.toml. I'm sorry @jromero but when there are so many `.`s in the filename I can no longer pay attention 🙃

Reply:
Based on today's WG, we have a few names to vote on from this [list](https://docs.google.com/document/d/1sq5TpJO-m8aTzgbAiVlS1QdNovne-3ku9FxxC2S_Zs4/edit#heading=h.4sikugimndzz).

Using emojis vote for your favorite:

🎉  = extensions
❤️ = stages
🚀 = initializers

Reply:
initializers is clear about the "when" (extensions and stages are ambiguous in this sense), and it also leaves room for hooks at the other parts of the lifecycle (e.g. finalizers)

Reply:
To me, "initializer" implies that the changes are scoped to initializing the build process. The changes applied the run image don't initialize anything.